File: python/Reverse Linked List.py
class ListNode1:
    def __init__(self, val):
        self.val = val
        self.next = None

    @staticmethod
    def reverseList(head):
        # Initialize prev pointer as NULL...
        prev = None
        # Initialize the curr pointer as the head...
        curr = head
        # Run a loop till curr points to NULL...
        while curr:
            # Initialize next pointer as the next pointer of curr...
            next_node = curr.next
            # Now assign the prev pointer to curr’s next pointer.
            curr.next = prev
            # Assign curr to prev, next to curr...
            prev = curr
            curr = next_node
        return prev


# Function to create a linked list from a list of values
def create_linked_list(values):
    if not values:
        return None
    head = ListNode1(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode1(val)
        current = current.next
    return head


# Function to print the linked list
def print_linked_list(head):
    while head:
        print(head.val, end=" -> ")
        head = head.next
    print("None")


# Example usage:
if __name__ == "__main__":
    # Input list
    input_list = [1, 2, 3, 4, 5]
    # Creating linked list
    head = create_linked_list(input_list)
    # Reversing the linked list
    reversed_head = ListNode1.reverseList(head)
    # Printing reversed linked list
    print("Reversed linked list:")
    print_linked_list(reversed_head)

'''
Given the head of a singly linked list, reverse the list, and return the reversed list.

 

Example 1:


Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]
Example 2:


Input: head = [1,2]
Output: [2,1]
Example 3:

Input: head = []
Output: []'''

File: JAVA/java/src/ReverseLinkedList.java

class ReverseLinkedList {
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }

    static class ReversedLinkedList {
        public ListNode reverseList(ListNode head) {
            ListNode prev = null;
            ListNode curr = head;
            while (curr != null) {
                ListNode nextNode = curr.next;
                curr.next = prev;
                prev = curr;
                curr = nextNode;
            }
            return prev;
        }

        // Function to create a linked list from an array of values
        public ListNode createLinkedList(int[] values) {
            if (values == null || values.length == 0) {
                return null;
            }
            ListNode head = new ListNode(values[0]);
            ListNode current = head;
            for (int i = 1; i < values.length; i++) {
                current.next = new ListNode(values[i]);
                current = current.next;
            }
            return head;
        }

        // Function to print the linked list
        public void printLinkedList(ListNode head) {
            while (head != null) {
                System.out.print(head.val + " -> ");
                head = head.next;
            }
            System.out.println("null");
        }
    }

    public static void main(String[] args) {
        ReversedLinkedList reversedLinkedList = new ReversedLinkedList();

        // Input array
        int[] inputArray = {1, 2, 3, 4, 5};

        // Creating linked list
        ListNode head = reversedLinkedList.createLinkedList(inputArray);

        // Reversing the linked list
        ListNode reversedHead = reversedLinkedList.reverseList(head);

        // Printing reversed linked list
        System.out.println("Reversed linked list:");
        reversedLinkedList.printLinkedList(reversedHead);
    }
}


